text/subsingleton-elimination.md

```diff
@@ -0,0 +1,176 @@
+    - any sigma-type of such types, as in `and`, possibly recursive as in `Acc`, possibly with indices as in `eq`. 
+ 
+These truncations are automatically considered as eliminable when the truncated type is syntactically recognized as belonging to the class (so-called "singleton elimination", which we could more accurately call "subsingleton elimination").
+
+The classes `Prop` and `SProp` are impredicative, in the sense that a dependent product of codomain `Prop` or `SProp` is again in the class.
+
+This CEPS is about three distinct extensions:
+- extending the definition of `Prop` and `SProp` classes with new syntactically decidable criteria
+- extending the dynamic detection of subsingleton elimination
+- adding a new class `HProp` providing an impredicative universe of subsingleton types; this extra `HProp` class would provide a variant of HoTT with impredicative `hProp` (`hProp := { A:Type & forall x y:A, x = y }`), alternative to using an axiom of resizing
+
+# Details of the current situation
+
+We review the current situation for `Prop` and `SProp`.
+
+## `Prop`
+
+Sort-polymorphism makes that inductive types are syntactically recognized in `Prop` after instantiation of parameters. For instance, `prod True True` is recognized as impredicative after instantiation:
+```
+Check Type -> prod True True.
+(* Type -> True * True
+     : Prop *)
+```
+Similarly in contexts where a `Prop` is expected:
+```
+Check and (prod True True) True
+(* True * True /\ True
+     : Prop *)
+```
+Contrastingly, subsingleton elimination is not recomputed dynamically:
+```
+Inductive PROD A B : Prop := PAIR : A -> B -> PROD A B.
+Check fun x : PROD True True => match x with PAIR _ _ _ _ => 0 end.
+(* Elimination of an inductive object of sort Prop is not allowed on a predicate in sort Set *)
+```
+which would actually be easy to address (see item 3 of the next section).
+
+## Actual `SProp`
+
+Actual `SProp` has no template polymorphism. Its subsingleton elimination is restricted to `False` so that dynamic computation of subsingleton elimination does not matter.
+
+We do not recommend template polymorphism for `SProp` (since template polymorphism is already planned to be superseded with universe polymorphism).
+
+In particular, we do not plan late detection of a parametric type in `Type` as an `SProp, nor late detection of a parametric type in `SProp` as supporting subsingleton type.
+
+# Detailed design for the `Prop` and `SProp` extensions
+
+## Uniformizing the treatment between `Prop` and `SProp`
+
+1. accept singleton types with all arguments in `SProp` to be in `SProp` as it is the case for `Prop`, the only difference being that recursive arguments are excluded; the indices should be in small types (i.e. `Prop` or `Set`)
+
+## Generalize the syntactic criterion for being in `Prop` or `SProp`
+ 
+2. accept in `SProp` and `Prop` singleton types with arguments of the form `forall a:A, B` when `B` is recursively an `SProp` or `Prop`-subsingleton
```

Comment:
Is it not already the case?

Reply:
You're right.
```
Inductive I : Prop := C : (nat->True) -> I.
```
is recognized as subsingleton.

I had naively checked
```
Inductive I : Prop := C : (nat->unit) -> I.
```
which is not recognized as such for other reasons.

Let me fix the document.